compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1458,40 +1490,75 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
         let all_impls = self.tcx.all_impls(trait_ref.def_id());
 
         match simp {
-            Some(simp) => all_impls
-                .filter_map(|def_id| {
-                    let imp = self.tcx.impl_trait_ref(def_id).unwrap();
-                    let imp_simp = fast_reject::simplify_type(
-                        self.tcx,
-                        imp.self_ty(),
-                        SimplifyParams::Yes,
-                        StripReferences::Yes,
-                    );
-                    if let Some(imp_simp) = imp_simp {
-                        if simp != imp_simp {
+            Some(simp) => {
+                all_impls
+                    .filter_map(|def_id| {
+                        if self.tcx.impl_polarity(def_id) == ty::ImplPolarity::Negative {
                             return None;
                         }
-                    }
-                    if self.tcx.impl_polarity(def_id) == ty::ImplPolarity::Negative {
-                        return None;
-                    }
-                    Some(imp)
-                })
-                .collect(),
+
+                        let imp = self.tcx.impl_trait_ref(def_id).unwrap();
+
+                        // Check for exact match.
+                        if trait_ref.skip_binder().self_ty() == imp.self_ty() {
+                            return Some(ImplCandidate {
+                                trait_ref: imp,
+                                similarity: CandidateSimilarity::Exact,
+                            });
+                        }
+
+                        // Check for match between simplified types.
+                        let imp_simp = fast_reject::simplify_type(
+                            self.tcx,
+                            imp.self_ty(),
+                            SimplifyParams::Yes,
+                            StripReferences::Yes,
+                        );
+                        if let Some(imp_simp) = imp_simp {
+                            if simp == imp_simp {
+                                return Some(ImplCandidate {
+                                    trait_ref: imp,
+                                    similarity: CandidateSimilarity::Simplified,
+                                });
+                            }
+                        }
+
+                        // Check for fuzzy match.
+                        // Pass `StripReferences::Yes` because although we do want to
+                        // be fuzzier than `simplify_type`, we don't want to be
+                        // *too* fuzzy.
```

Comment:
that reasoning seems unclear to me :thinking: how does passing `StripReferences::Yes` prevent ourselves from being "*too* fuzzy"?